### PR TITLE
Enable MSI for deployments.

### PR DESCRIPTION
**Pull request by kwoodson**
Investigating enabling MSI for OSA deployments.   Very WIP.

### CODE REVIEW

**Comment by kwoodson** [comment]
Placing this value in the `/etc/origin/cloudprovider/azure.conf` was able to get the following message:
```
useManagedIdentityExtension: true
```
```
    4882 azure_auth.go:59] azure: using managed identity extension to retrieve access token
```

File: pkg/arm/data/master-startup.sh [hunk 1]
```diff
@@ -396,10 +396,9 @@ mkdir -p /etc/origin/cloudprovider
 
 # TODO: this is duplicated, and that's not ideal
 cat >/etc/origin/cloudprovider/azure.conf <<'EOF'
+useManagedIdentityExtension: true
 tenantId: {{ .Config.TenantID | quote }}
 subscriptionId: {{ .Config.SubscriptionID | quote }}
-aadClientId: {{ .ContainerService.Properties.ServicePrincipalProfile.ClientID | quote }}
-aadClientSecret: {{ .ContainerService.Properties.ServicePrincipalProfile.Secret | quote }}
 aadTenantId: {{ .Config.TenantID | quote }}
```
Comment (kwoodson) [inline pkg/arm/data/master-startup.sh]: Is this field also one we can remove?

**Comment by kwoodson** [comment]
Controllers appear to be in `CrashLoopBackOff`:
```
kube-system   controllers-master-000000   0/1       CrashLoopBackOff   48         3h
kube-system   controllers-master-000001   0/1       CrashLoopBackOff   48         3h
kube-system   controllers-master-000002   0/1       CrashLoopBackOff   48         3h
```

```
F0802 00:47:01.394144       1 controllermanager.go:166] error building controller context: cloud provider could not be initialized: could not init cloud provider "azure": Getting the managed service identity endpoint: open /var/lib/waagent/ManagedIdentity-Settings: no such file or directory
```

Seems like this is related:
https://github.com/kubernetes/kubernetes/issues/61456

I attempted to fix this by adding a volume mount `/var/lib/waagent/ManagedIdentity-Settings` to the controllers.yaml and it appeared to work.

```
    - mountPath: /var/lib/waagent/ManagedIdentity-Settings
      name: master-msi
      readOnly: true
...
  - hostPath:
      path: /var/lib/waagent/ManagedIdentity-Settings
    name: master-msi
```

Docker logs:
```
{"log":"I0802 01:08:55.800486       1 azure_auth.go:59] azure: using managed identity extension to retrieve access token\n","stream":"stderr","time":"2018-08-02T01:08:55.801175698Z"}
```

**Comment by kwoodson** [comment]
Latest deploy failed with:
```
{
  "error": {
    "code": "RoleAssignmentUpdateNotPermitted",
    "message": "Tenant ID, application ID, principal ID, and scope are not allowed to be updated."
  }
}
```

This was a fresh deployment into a new resource group.  Need to investigate further.

File: pkg/arm/data/master-startup.sh (hunk 1 above)
Comment (kargakis) [reply pkg/arm/data/master-startup.sh]: @sozercan

**Comment by kargakis** [comment]
[The docker registry is also using azure.conf for setting up its storage account](https://github.com/openshift/openshift-azure/blob/edb66ebe3a7e82f8810342b9872a45971aa24ba5/pkg/addons/data/Secret/default/etc-origin-cloudprovider.yaml). I guess we need to set `useManagedIdentity: true` there, too?

**Comment by kwoodson** [comment]
I have spoken with Certac and he believes we are hitting this issue https://github.com/Azure/acs-engine/issues/2916

Looks like a bug or issue with removing old roleAssignments.  When the VM gets removed it needs to cascade and remove the roleAssignments as well.  Seems though this isn't happening properly.

**Comment by kwoodson** [comment]
@kargakis, I'm not sure the docker-registry supports MSI.  It could possibly but the container would have to understand auth tokens.  This is the documentation on the registry:
https://github.com/docker/docker.github.io/blob/master/registry/storage-drivers/azure.md

There is no mention about MSI.

**Comment by kargakis** [comment]
> @kargakis, I'm not sure the docker-registry supports MSI. It could possibly but the container would have to understand auth tokens. This is the documentation on the registry:
> https://github.com/docker/docker.github.io/blob/master/registry/storage-drivers/azure.md
> 
> There is no mention about MSI.

For clarity, as discussed in scrum, the docker registry does not need to use MSI, it's the init container that needs access to get the storage account secret.

**Comment by kwoodson** [comment]
Latest test with fixes for the `master-config.yaml` template.
```
Pushing image docker-registry.default.svc:5000/test/cakephp-mysql-persistent:latest ...
Pushed 0/10 layers, 0% complete
Pushed 1/10 layers, 13% complete
Pushed 2/10 layers, 20% complete
Push successful
```

**Comment by kwoodson** [comment]
@kargakis, should this build pass the ci/prow/build?  It looks like its failing on the `@generated` bits:
```
+ diff -r /go/src/github.com/openshift/openshift-azure/hack/../pkg/addons/bindata.go /tmp/tmp.hGWDMlWCfV/pkg/addons/bindata.go
1c1
< // Code generated by go-bindata. DO NOT EDIT.
---
> // Code generated by go-bindata. DO NOT EDIT. @generated
+ echo 'go generate produced changes that were not already present'
+ exit 1
go generate produced changes that were not already present
+ rm -rf /tmp/tmp.hGWDMlWCfV
make: *** [verify] Error 1
```

Any suggestion?

**Comment by kargakis** [comment]
You need to update your go-bindata binary

go get -u github.com/go-bindata/go-bindata

**Comment by kwoodson** [comment]
@kargakis, thanks.  I've updated and amended.  For my own sake, is this something that we need to do often or a 1-off?

**Comment by kargakis** [comment]
1-off. Will review tomorrow

**Comment by kwoodson** [comment]
Updated my local `go-bindata` pkg and re-generated.

**Comment by kwoodson** [comment]
@kargakis, It currently binds in the cloudprovider file but the credentials are not used there.  The credentials in the master-config.yaml are required for aad integration.  Those are the only credentials now.

**Comment by kwoodson** [comment]
I have a proposed PR to replace this one using the azure-sdk-for-go https://github.com/openshift/openshift-azure/pull/170/

File: pkg/addons/translate.go [hunk 2]
```diff
@@ -316,31 +316,6 @@ var Translations = map[string][]struct {
...
-			NestedPath: jsonpath.MustCompile("$.aadClientSecret"),
-			Template:   "{{ .ContainerService.Properties.ServicePrincipalProfile.Secret }}",
-		},
-		{
-			Path:       jsonpath.MustCompile("$.stringData.'azure.conf'"),
-			NestedPath: jsonpath.MustCompile("$.aadTenantId"),
-			Template:   "{{ .Config.TenantID }}",
-		},
 		{
 			Path:       jsonpath.MustCompile("$.stringData.'azure.conf'"),
```
Comment (pweil-) [inline pkg/addons/translate.go]: still need a secret here or is a config map more appropriate now?  Not a blocker

File: pkg/arm/data/azuredeploy.json [hunk 3]
```diff
@@ -163,6 +163,20 @@
                 ]
             }
         },
+        {
+            "apiVersion": "2014-10-01-preview",
+            "name": "[guid(concat('Microsoft.Compute/virtualMachineScaleSets/', 'ss-{{ $app.Name }}', 'vmidentity'))]",
+            "type": "Microsoft.Authorization/roleAssignments",
+            "properties": {
+{{- if or (eq $app.Role "master") (eq $app.Role "infra") }}
+              "roleDefinitionId": "[concat('/subscriptions/', subscription().subscriptionId, '/providers/Microsoft.Authorization/roleDefinitions/', 'b24988ac-6180-42a0-ab88-20f7382dd24c')]",
```
Comment (pweil-) [inline pkg/arm/data/azuredeploy.json]: magic numbers! :tada:

File: hack/ssh.sh [hunk 4]
```diff
@@ -11,5 +11,9 @@ ID=$3
 
 IP=$(az vmss list-instance-public-ips -g $RESOURCEGROUP -n $SS --query "[$ID].ipAddress" | tr -d '"')
 
+scp -o UserKnownHostsFile=/dev/null -o StrictHostKeyChecking=no \
```
Comment (pweil-) [inline hack/ssh.sh]: Note: unrelated to MSI - this is to allow you to ssh from master to other nodes. It tripped me up :smile: 

@kwoodson might just want to put this in its own commit or PR

**Comment by pweil-** [comment]
@sozercan bump for review

File: hack/ssh.sh (hunk 4 above)
Comment (kwoodson) [reply hack/ssh.sh]: @pweil-, ok, i can do that.

File: pkg/arm/data/azuredeploy.json (hunk 3 above)
Comment (kwoodson) [reply pkg/arm/data/azuredeploy.json]: Yeah, not sure what to do here as these were given to me.  They are hard coded.

**Comment by sozercan** [comment]
sorry about the delay reviewing this, we are testing MSI to make sure it works in managed apps
